assets/app/scripts/services/navigate.js

```diff
@@ -46,6 +46,60 @@ angular.module("openshiftConsole")
        */
       toNextSteps: function(name, projectName){
         $location.path("project/" + encodeURIComponent(projectName) + "/create/next").search("name", name);
+      },
+
+      // Resource is either a resource object, or a name.  If resource is a name, kind and namespace must be specified
+      // Note that builds and deployments can only have their URL built correctly (including their config in the URL)
+      // if resource is an object, otherwise they will fall back to the non-nested URL.
+      resourceURL: function(resource, kind, namespace) {
+        if (!resource || (!resource.metadata && (!kind || !namespace))) {
+          return null;
+        }
+
+        // normalize based on the kind of args we got
+        if (!kind) {
+          kind = resource.kind;
+        }
+        if (!namespace) {
+          namespace = resource.metadata.namespace; 
+        }
+        var name = resource;
+        if (resource.metadata) {
+          name = resource.metadata.name;
+        }
+
+        var url = "project/" + namespace + "/browse/";
+        switch(kind) {
+          case "Build":
+            if (resource.metadata && resource.metadata.labels && resource.metadata.labels.buildconfig) {
+              url += "builds/" + resource.metadata.labels.buildconfig + "/" + name;
+            }
+            else {
+              url += "builds-noconfig/" + name;
+            }
+            break;
+          case "BuildConfig":
+            url += "builds/" + name;
+            break;
+          case "DeploymentConfig":
+            url += "deployments/" + name;
+            break;
+          case "ReplicationController":
+            var depConfig = resource.metadata ? annotationFilter(resource, 'deploymentConfig') : null;
+            if (depConfig) {
+              url += "deployments/" + depConfig + "/" + name;
```

Comment:
I feel like we should be calling `encodeURIComponent` on these path segments to be safe in case we ever get bad data, even if name is only supposed to have valid characters in a URL. I'm not sure on the rule for annotations.